### ISSUE TITLE
step_method_dict raise an error

### ISSUE THREAD

**Issue by isofer** [pymc-devs/pymc3#32]
(the markdown of the following code misses a few lines)

I'm running the following:

```
from pymc import *
from numpy import *
from numpy.random import rand


def bla():

    t = 50
    data = rand(t)
    e = 2
    k = 5;
    R = eye(k);
    D = rand(t,3)
    Xdm = rand(t,k,e)

    h = MvNormal('h', zeros(k), eye(k))

    alpha = Uniform('alpha', np.ones(e) * 0, np.ones(e) * 100)

    gamma = MvNormal('gamma', zeros(3), eye(3))

    @deterministic
    def drift(D = D, gamma=gamma):
        return dot(D, gamma)

    sigma = InverseGamma('sigma', alpha=4, beta=4, value=10)

    @deterministic
    def all_alphaXh(Xdm=Xdm, h=h, alpha=alpha, t=t):
        signal = np.zeros(t)
        for i in range(len(alpha)):
            signal += np.dot(Xdm[:,:,i], h) * alpha[i]

        return  signal;

    y = Normal('y', all_alphaXh + drift, 1/sigma, value=data, observed=True)
    return { 'h':h, 'alpha':alpha, 'gamma':gamma, 'sigma':sigma,
             'y':y, 'drift':drift, 'all_alphaXh':all_alphaXh}

m = MCMC(bla())



an when I try to access m.step_method_dict I get the following error

In [3]: m.step_method_dict
Out[3]: ---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)

/Users/imrisofer/<ipython console> in <module>()

/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/site-packages/IPython/Prompts.pyc in __call__(self, arg)
    550 
    551             # and now call a possibly user-defined print mechanism

--> 552             manipulated_val = self.display(arg)
    553 
    554             # user display hooks can change the variable to be stored in


/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/site-packages/IPython/Prompts.pyc in _display(self, arg)
    576             return IPython.generics.result_display(arg)
    577         except TryNext:
--> 578             return self.shell.hooks.result_display(arg)
    579 
    580     # Assign the default display method:


/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/site-packages/IPython/hooks.pyc in __call__(self, *args, **kw)
    139             #print "prio",prio,"cmd",cmd #dbg

    140             try:
--> 141                 ret = cmd(*args, **kw)
    142                 return ret
    143             except ipapi.TryNext, exc:

/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/site-packages/IPython/hooks.pyc in result_display(self, arg)
    169 
    170     if self.rc.pprint:
--> 171         out = pformat(arg)
    172         if '\n' in out:
    173             # So that multi-line strings line up with the left column of


/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/pprint.pyc in pformat(self, object)
    117     def pformat(self, object):
    118         sio = _StringIO()
--> 119         self._format(object, sio, 0, 0, {}, 0)
    120         return sio.getvalue()
    121 

/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/pprint.pyc in _format(self, object, stream, indent, allowance, context, level)
    135             self._readable = False
    136             return
--> 137         rep = self._repr(object, context, level - 1)
    138         typ = _type(object)
    139         sepLines = _len(rep) > (self._width - 1 - indent - allowance)

/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/pprint.pyc in _repr(self, object, context, level)
    228     def _repr(self, object, context, level):
    229         repr, readable, recursive = self.format(object, context.copy(),
--> 230                                                 self._depth, level)
    231         if not readable:
    232             self._readable = False

/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/pprint.pyc in format(self, object, context, maxlevels, level)
    240         and whether the object represents a recursive construct.
    241         """
--> 242         return _safe_repr(object, context, maxlevels, level)
    243 
    244 

/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/pprint.pyc in _safe_repr(object, context, maxlevels, level)
    282         level += 1
    283         saferepr = _safe_repr
--> 284         for k, v in _sorted(object.items()):
    285             krepr, kreadable, krecur = saferepr(k, context, maxlevels, level)
    286             vrepr, vreadable, vrecur = saferepr(v, context, maxlevels, level)

/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/pprint.pyc in _sorted(iterable)
     77             warnings.filterwarnings("ignore", "comparing unequal types "
     78                                     "not supported", DeprecationWarning)
---> 79         return sorted(iterable)
     80 
     81 class PrettyPrinter:

/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/site-packages/pymc/CommonDeterministics.pyc in new_method(self, other)
    631                                 {'self':self, 'other':other, 'op':'__'+op_name+'__'},
    632                                 trace=False,
--> 633                                 plot=False)
    634     # Convert the function into a method for klass.

    635     new_method.__name__ = '__'+op_name+'__'

/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/site-packages/pymc/PyMCObjects.pyc in __init__(self, eval, doc, name, parents, dtype, trace, cache_depth, plot, verbose, jacobians, jacobian_formats)
    396                         trace=trace,
    397                         plot=plot,
--> 398                         verbose=verbose)
    399 
    400         # self._value.force_compute()


/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/site-packages/pymc/Node.pyc in __init__(self, doc, name, parents, cache_depth, trace, dtype, plot, verbose)
    218         self.extended_children = set()
    219 
--> 220         Node.__init__(self, doc, name, parents, cache_depth, verbose=verbose)
    221 
    222         if self.dtype is None:

/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/site-packages/pymc/Node.pyc in __init__(self, doc, name, parents, cache_depth, verbose)
    136 
    137         # Initialize

--> 138         self.parents = parents
    139 
    140     def _get_parents(self):

/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/site-packages/pymc/Node.pyc in _set_parents(self, new_parents)
    157 
    158         # Get new lazy function

--> 159         self.gen_lazy_function()
    160 
    161     parents = property(_get_parents, _set_parents, doc="Self's parents: the variables referred to in self's declaration.")

/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/site-packages/pymc/PyMCObjects.pyc in gen_lazy_function(self)
    410                                     cache_depth = self._cache_depth)
    411 
--> 412         self._value.force_compute()
    413 
    414         self._jacobians = {}

/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/site-packages/pymc/LazyFunction.so in pymc.LazyFunction.LazyFunction.force_compute()

/Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/site-packages/pymc/CommonDeterministics.pyc in eval_fun(self, other, op)
    625         # This code creates a Deterministic object.

    626         def eval_fun(self, other, op):
--> 627             return getattr(self, op)(other)
    628         return pm.Deterministic(eval_fun,
    629                                 'A Deterministic returning the value of %s(%s,%s)'%(op_name,self.__name__, str(other)),

ValueError: shape mismatch: objects cannot be broadcast to a single shape
```

in order to reproduce this bug one need to type m.step_method_dict in the interpreter. execute the same command from a file does not raise an error.

using Mac OS 10.6.7, pymc 2.2grad, python 2.7.1

**Comment by fonnesbeck** [comment]
I cannot replicate this on pymc 2.2 alpha and python 2.6.1:

```
In [3]: m.step_method_dict
Out[3]: 
{<pymc.distributions.MvNormal 'h' at 0x104dae490>: [],
 <pymc.distributions.Uniform 'alpha' at 0x107b0cb50>: [],
 <pymc.distributions.MvNormal 'gamma' at 0x107b0cb90>: [],
 <pymc.distributions.InverseGamma 'sigma' at 0x107b0cc10>: []}
```

So its a matter of tracking down whether its due to your Python distribution or the gradient branch.

**Comment by isofer** [comment]
I'm using the python that came with the latest EPD package.
Do you get this error when using pymc 2.2grad?

**Comment by apatil** [comment]
I get that too. It's because IPython likes to sort things before displaying them. 

This part of the stack trace:

``` python
> /Library/Frameworks/EPD64.framework/Versions/7.0/lib/python2.7/pprint.py(79)_sorted()
     78                                     "not supported", DeprecationWarning)
---> 79         return sorted(iterable)
     80 
```

tries to sort 

``` python
[(<pymc.distributions.MvNormal 'h' at 0x10ca52210>, []), (<pymc.distributions.Uniform 'alpha' at 0x10f42ec90>, []), (<pymc.distributions.MvNormal 'gamma' at 0x10f42ecd0>, []), (<pymc.distributions.InverseGamma 'sigma' at 0x10f42ed50>, [])]
```

which ends up trying to sort using the `__lt__` method of Variable, which produces another Variable.

This should probably be reported as a bug in IPython... if it doesn't manage to sort things, it should just display them unsorted.

Anand
which

**Comment by apatil** [comment]
See https://github.com/ipython/ipython/issues/417

**Comment by fonnesbeck** [comment]
Can anyone replicate this in the just-released iPython 0.11?
